awx/main/tasks.py

```diff
@@ -2805,6 +2806,8 @@ def emit(self, record):
             # for us.
             save_status, tb, exc = cmd.perform_update(options, data, inventory_update)
         except Exception as raw_exc:
+            if exc is None:
```

Comment:
is it even possible for `perform_update` to return any values if it threw an exception? wondering if `exc` should be dropped entirely in the except clause and replaced with `raw_exc`. I'm fuzzy on the context here, though, so not sure that I'm reading this correctly.

In any case, sounds like maybe you caught the root cause of some host limit failures, which is awesome.

Reply:
Oh yeah, you might have a point, I knew this was ugly, but maybe it's just not even correct given that.

If you read the code of the method, it's a mess. A license error can result in multiple different scenarios. I could throw and uncaught `CommandError`, or it could use the internal license error exception and be caught within the method. If the method catches it, then it won't throw an exception here, but `exc` will be non-null.

We should refactor that if possible.

One thing to be mindful of - our behavior should be the same for all license error scenarios. That's going to be kind of hard, because if it's a `CommandError`, then how do you know if it was a license failure? If refactoring, you might start with that, and maybe enforce that `perform_update` _never_ throws a `CommandError`, and replace those cases with license error? I end with a question mark, because this is speculative, and it needs more thought and consideration.